drivers/digital-io/max14906/max14906.c

```diff
@@ -0,0 +1,407 @@
+
+	memset(desc->buff, 0, xfer.bytes_number);
+	desc->buff[0] = no_os_field_prep(MAX14906_CHIP_ADDR_MASK, desc->chip_address);
+	desc->buff[0] |= no_os_field_prep(MAX14906_ADDR_MASK, addr);
+	desc->buff[0] |= no_os_field_prep(MAX14906_RW_MASK, 0);
+
+	if (desc->crc_en)
+		desc->buff[2] = max14906_crc(&desc->buff[0], true);
+
+	ret = no_os_spi_transfer(desc->comm_desc, &xfer, 1);
+	if (ret)
+		return ret;
+
+	if (desc->crc_en) {
+		crc = max14906_crc(&desc->buff[0], false);
+		if (crc != desc->buff[2])
+			return -EINVAL;
+	}
+
+	*val = desc->buff[1];
+
+	return 0;
+}
+
+/**
+ * @brief Update the value of a device register (read/write sequence).
+ * @param desc - device descriptor for the MAX14906
+ * @param addr - address of the register
+ * @param mask - bit mask of the field to be updated
+ * @param val - value of the masked field. Should be bit shifted by using
+ * 		 no_os_field_prep(mask, val)
+ * @return 0 in case of success, negative error code otherwise
+ */
+int max14906_reg_update(struct max14906_desc *desc, uint32_t addr,
+			uint32_t mask, uint32_t val)
+{
+	int ret;
+	uint32_t reg_val = 0;
+
+	ret = max14906_reg_read(desc, addr, &reg_val);
+	if (ret)
+		return ret;
+
+	reg_val &= ~mask;
+	reg_val |= mask & val;
+
+	return max14906_reg_write(desc, addr, reg_val);
+}
+
+/**
+ * @brief Read the (voltage) state of a channel (works for both input or output).
+ * @param desc - device descriptor for the MAX14906
+ * @param ch - channel index (0 based).
+ * @param val - binary value representing a channel's voltage level (0 or 1).
+ * @return 0 in case of success, negative error code otherwise
+ */
+int max14906_ch_get(struct max14906_desc *desc, uint32_t ch, uint32_t *val)
+{
+	int ret;
+
+	if (ch >= MAX14906_CHANNELS)
+		return -EINVAL;
+
+	ret = max14906_reg_read(desc, MAX14906_DOILEVEL_REG, val);
+	if (ret)
+		return ret;
+
+	*val = no_os_field_get(MAX14906_DOI_LEVEL_MASK(ch), *val);
+
+	return 0;
+}
+
+/**
+ * @brief Write the (voltage) state of a channel (only for output channels).
+ * @param desc - device descriptor for the MAX14906
+ * @param ch - channel index (0 based).
+ * @param val - binary value representing a channel's voltage level (0 or 1).
+ * @return 0 in case of success, negative error code otherwise
+ */
+int max14906_ch_set(struct max14906_desc *desc, uint32_t ch, uint32_t val)
+{
+	if (ch >= MAX14906_CHANNELS)
+		return -EINVAL;
+
+	return max14906_reg_update(desc, MAX14906_SETOUT_REG,
+				   MAX14906_HIGHO_MASK(ch), (val) ?
+				   MAX14906_HIGHO_MASK(ch) : 0);
+}
+
+/**
+ * @brief Configure a channel's function.
+ * @param desc - device descriptor for the MAX14906
+ * @param ch - channel index (0 based).
+ * @param function - channel configuration (input, output or high-z).
+ * @return 0 in case of success, negative error code otherwise
+ */
+int max14906_ch_func(struct max14906_desc *desc, uint32_t ch,
+		     enum max14906_function function)
+{
+	int ret;
+
+	if (function == MAX14906_HIGH_Z) {
```

Comment:
use `swich ... case` and variables to set the contents of MAX14906_CONFIG_DO_REG register and MAX14906_SETOUT_REG. Then call the reg_update functions only once for each register.